### ISSUE TITLE
App falls down (out of memory) when 1 node is using virtual network interface

### ISSUE THREAD

**Issue by BartosStore** [hazelcast/hazelcast#13035]
Hi,
  we are using hazelcast for synchronization between 3 nodes (3 PCs) of our application. 

  Node1 and node2 use standard network interface. Node3 uses virtual network interface (for example virtualbox interface). Node3 fell down with exception. So, we disabled every network interface except standard interface. We started node3 again. Then node2 fell down with same exception. It fell down with out of memory exception cause there was byte array with 1,8 GB.

  We are using hazelcast 3.9.2.

  Here is log from server:

```
2018-05-03 10:19:20.695 +02:00 INFO  [hz._hzInstance_1_Scaler-2bb3dd27-hzl-cluster}.cached.thread-5] com.hazelcast.nio.tcp.TcpIpConnector - [192.168.51.53]:5701 [Scaler-2bb3dd27-hzl-cluster}] [3.9.2] Connecting to /192.168.99.1:5701, timeout: 0, bind-any: true
2018-05-03 10:19:20.696 +02:00 INFO  [hz._hzInstance_1_Scaler-2bb3dd27-hzl-cluster}.cached.thread-5] com.hazelcast.nio.tcp.TcpIpConnectionManager - [192.168.51.53]:5701 [Scaler-2bb3dd27-hzl-cluster}] [3.9.2] Established socket connection between /192.168.99.1:51501 and /192.168.99.1:5701
2018-05-03 10:19:20.696 +02:00 INFO  [hz._hzInstance_1_Scaler-2bb3dd27-hzl-cluster}.IO.thread-Acceptor] com.hazelcast.nio.tcp.TcpIpAcceptor - [192.168.51.53]:5701 [Scaler-2bb3dd27-hzl-cluster}] [3.9.2] Accepting socket connection from /192.168.99.1:51501
2018-05-03 10:19:20.696 +02:00 INFO  [hz._hzInstance_1_Scaler-2bb3dd27-hzl-cluster}.cached.thread-3] com.hazelcast.nio.tcp.TcpIpConnectionManager - [192.168.51.53]:5701 [Scaler-2bb3dd27-hzl-cluster}] [3.9.2] Established socket connection between /192.168.99.1:5701 and /192.168.99.1:51501
2018-05-03 10:19:20.698 +02:00 WARN  [hz._hzInstance_1_Scaler-2bb3dd27-hzl-cluster}.IO.thread-in-1] com.hazelcast.nio.tcp.TcpIpConnectionManager - [192.168.51.53]:5701 [Scaler-2bb3dd27-hzl-cluster}] [3.9.2] Wrong bind request from [192.168.51.53]:5701! This node is not the requested endpoint: [192.168.99.1]:5701
2018-05-03 10:19:20.698 +02:00 INFO  [hz._hzInstance_1_Scaler-2bb3dd27-hzl-cluster}.IO.thread-in-1] com.hazelcast.nio.tcp.TcpIpConnection - [192.168.51.53]:5701 [Scaler-2bb3dd27-hzl-cluster}] [3.9.2] Connection[id=65, /192.168.99.1:5701->/192.168.99.1:51501, endpoint=null, alive=false, type=MEMBER] closed. Reason: Wrong bind request from [192.168.51.53]:5701! This node is not the requested endpoint: [192.168.99.1]:5701
2018-05-03 10:19:20.699 +02:00 INFO  [hz._hzInstance_1_Scaler-2bb3dd27-hzl-cluster}.IO.thread-in-0] com.hazelcast.nio.tcp.TcpIpConnection - [192.168.51.53]:5701 [Scaler-2bb3dd27-hzl-cluster}] [3.9.2] Connection[id=64, /192.168.99.1:51501->/192.168.99.1:5701, endpoint=[192.168.99.1]:5701, alive=false, type=MEMBER] closed. Reason: Connection closed by the other side
2018-05-03 10:19:20.700 +02:00 INFO  [hz._hzInstance_1_Scaler-2bb3dd27-hzl-cluster}.cached.thread-3] com.hazelcast.nio.tcp.TcpIpConnector - [192.168.51.53]:5701 [Scaler-2bb3dd27-hzl-cluster}] [3.9.2] Connecting to /192.168.99.1:5701, timeout: 0, bind-any: true
2018-05-03 10:19:20.700 +02:00 INFO  [hz._hzInstance_1_Scaler-2bb3dd27-hzl-cluster}.cached.thread-3] com.hazelcast.nio.tcp.TcpIpConnectionManager - [192.168.51.53]:5701 [Scaler-2bb3dd27-hzl-cluster}] [3.9.2] Established socket connection between /192.168.99.1:51502 and /192.168.99.1:5701
2018-05-03 10:19:20.701 +02:00 INFO  [hz._hzInstance_1_Scaler-2bb3dd27-hzl-cluster}.IO.thread-Acceptor] com.hazelcast.nio.tcp.TcpIpAcceptor - [192.168.51.53]:5701 [Scaler-2bb3dd27-hzl-cluster}] [3.9.2] Accepting socket connection from /192.168.99.1:51502
2018-05-03 10:19:20.701 +02:00 INFO  [hz._hzInstance_1_Scaler-2bb3dd27-hzl-cluster}.cached.thread-5] com.hazelcast.nio.tcp.TcpIpConnectionManager - [192.168.51.53]:5701 [Scaler-2bb3dd27-hzl-cluster}] [3.9.2] Established socket connection between /192.168.99.1:5701 and /192.168.99.1:51502
2018-05-03 10:19:20.701 +02:00 WARN  [hz._hzInstance_1_Scaler-2bb3dd27-hzl-cluster}.IO.thread-in-0] com.hazelcast.nio.tcp.TcpIpConnectionManager - [192.168.51.53]:5701 [Scaler-2bb3dd27-hzl-cluster}] [3.9.2] Wrong bind request from [192.168.51.53]:5701! This node is not the requested endpoint: [192.168.99.1]:5701
2018-05-03 10:19:20.701 +02:00 INFO  [hz._hzInstance_1_Scaler-2bb3dd27-hzl-cluster}.IO.thread-in-0] com.hazelcast.nio.tcp.TcpIpConnection - [192.168.51.53]:5701 [Scaler-2bb3dd27-hzl-cluster}] [3.9.2] Connection[id=67, /192.168.99.1:5701->/192.168.99.1:51502, endpoint=null, alive=false, type=MEMBER] closed. Reason: Wrong bind request from [192.168.51.53]:5701! This node is not the requested endpoint: [192.168.99.1]:5701
2018-05-03 10:19:20.701 +02:00 INFO  [hz._hzInstance_1_Scaler-2bb3dd27-hzl-cluster}.IO.thread-in-2] com.hazelcast.nio.tcp.TcpIpConnection - [192.168.51.53]:5701 [Scaler-2bb3dd27-hzl-cluster}] [3.9.2] Connection[id=66, /192.168.99.1:51502->/192.168.99.1:5701, endpoint=[192.168.99.1]:5701, alive=false, type=MEMBER] closed. Reason: Connection closed by the other side
```

  Memory usage:

![graph](https://user-images.githubusercontent.com/6641915/39586433-4ff052c4-4ef7-11e8-869a-16bbe7af3546.PNG)

  Class tree:

![tree](https://user-images.githubusercontent.com/6641915/39586477-72a27fe0-4ef7-11e8-8200-e01257cfa486.PNG)

  Values of byte[]:

![byte_array](https://user-images.githubusercontent.com/6641915/39586510-8525a78c-4ef7-11e8-9bbb-51bd681101ab.PNG)

  It seems that there is no cleaner for connections.

**Comment by pveentjer** [comment]
Thanks for the excellent bug report.

**Comment by pveentjer** [comment]
I think there are 2 problems:

1: the actual connection eventually being rejected with the message ' This node is not the requested endpoint'. 

2: the IO-Balancer not getting rid of such rejected connections. Which eventually leads to an OOME.

**Comment by pveentjer** [comment]
Can you flip on diagnostics? 

http://docs.hazelcast.org/docs/latest-development/manual/html/Management/Diagnostics/Enabling_Diagnostics_Logging.html

Quite useful to diagnose problems. The above configuration has very little overhead; I use it when doing benchmarks.

**Comment by pveentjer** [comment]
I think I have found the cause of the problem. The issue is that the connection isn't removed from the NioEventLoopGroup and then it isn't removed from the IOBalancer.

The reason it isn't removed from the NioEventLoopGroup is that the ConnectionListener doesn't work because of a check:

```
@Override
    public void onConnectionClose(Connection connection) {
        closedCount.inc();

        activeConnections.remove(connection);

        Address endPoint = connection.getEndPoint();
        if (endPoint != null) {<------------
            connectionsInProgress.remove(endPoint);
            connectionsMap.remove(endPoint, connection);
            fireConnectionRemovedEvent(connection, endPoint);
        }
    }
```

So the endpoint needs to be set for the listeners to work.

However the endpoint is only set after the bind request validation, which in your case is failing.
```
// Prevent BINDs that don't have this node as the destination
        if (!ensureValidBindTarget(connection, remoteEndPoint, localEndpoint, thisAddress)) {
            return false;
        }

        connection.setEndPoint(remoteEndPoint);<---- endpoint set
        ioService.onSuccessfulConnection(remoteEndPoint);
        if (reply) {
            sendBindRequest(connection, remoteEndPoint, false);
        }
```

So this leads to the ConnectionListener not being called. Therefor the Channel isn't removed from the NioEventLoopGroup and therefor the Inbound/Outbound pipelines aren't removed from the IOBalancer. And this leads to your memory leak.

**Comment by pveentjer** [comment]
My guess is that there is some kind of NAT going on that is causing problems. The IP address inside the virtualbox is different than the outside IP address to connect to it. 

Perhaps you can try bridged networking:
https://www.virtualbox.org/manual/ch06.html#network_bridged

**Comment by mmedenjak** [comment]
We've had the error `This node is not the requested endpoint` many times while using docker and AWS and in all cases it was because of NAT - the member was binding to one address while other members/clients are connecting to a different address.
The workarounds are specifying the public address (http://docs.hazelcast.org/docs/latest/manual/html-single/#public-address) or implementing a member address provider (http://docs.hazelcast.org/docs/latest/manual/html-single/#member-address-provides-spi)

**Comment by mmedenjak** [comment]
@BartosStore can you try upgrading to Hazelcast 3.10 to avoid the OOME or use the public address and member address provider to avoid the connection issues?

**Comment by BartosStore** [comment]
We upgraded Hazelcast to 3.10 and issue will be verified soon. 

Thank you for your help.

**Comment by hos-p** [comment]
Hi,
I will reply instead of @BartosStore. We have upgraded to Hazelcast 3.10 and made the same configuration. The issue with OOM persists.

![shortest_path](https://user-images.githubusercontent.com/6361016/41056893-d9ee39da-69c5-11e8-984b-70be1f4775a5.PNG)

![histogram](https://user-images.githubusercontent.com/6361016/41056901-e0eaea44-69c5-11e8-8413-f66ef504382c.PNG)

![bytes](https://user-images.githubusercontent.com/6361016/41056905-e65a995c-69c5-11e8-8ef2-0585d88ecc32.PNG)

**Comment by mmedenjak** [comment]
Hi guys, sorry for the delay. I have tried simulating this scenario and even though I get the same error, there is no memory leak.

Even though I see that the `IOBalancer` is consuming all of the heap, from your screenshot I cannot see which objects inside the balancer are retaining the heap. Can you post the heap dump here?

@pveentjer I don't think the problem is in the `connectionsInProgress`. I have been observing its state and it doesn't grow. Also, it is a `Set<Address>` which means we would need to add different addresses each time.

**Comment by mmedenjak** [comment]
Closing as there is no response from the user. We have made significant changes in the IO subsystem since 3.10. 
There have been significant changes regarding the `This node is not the requested endpoint` and the `Wrong bind request` errors in this PR. The entire `TcpIpConnectionManager` has been removed and replaced with the `TcpIpEndpointManager` which doesn't do these spoofing checks anymore.

Since we are unable to reproduce the issue, we need a heap dump to analyse where objects are being retained. @hos-p @BartosStore feel free to ping us again with the heap dump if possible.
